### PR TITLE
API Versioning

### PR DESCRIPTION
**Pull request by codingupastorm**
- The existing API can be considered "V1" of the API. All of the URLs remain as they are so no issues for the wallet etc.

- A v2.0-dev API has been created that we can add to. It uses `v2.0-dev` as a URL path parameter (in real-life this will be `v2` so paths will be `api/v2/smartcontracts/code`. It's currently empty save for a demo call.

- Swagger automatically groups the controllers for and renders both, and each can be selected to view via dropdown.

See [gif of swagger](https://media.giphy.com/media/ch8HV38RJT2Y8Krug0/giphy.gif)

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.SmartContracts/ReflectionExecutor/Controllers/SmartContractsController.cs [hunk 1]
```diff
@@ -28,10 +28,10 @@
 using Stratis.SmartContracts.Core;
 using Stratis.SmartContracts.Core.Receipts;
 using Stratis.SmartContracts.Core.State;
-using State = Stratis.SmartContracts.CLR.State;
 
 namespace Stratis.Bitcoin.Features.SmartContracts.ReflectionExecutor.Controllers
 {
+    [ApiVersion("1.0")]
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Features.SmartContracts/ReflectionExecutor/Controllers/SmartContractsController.cs]: Is the choice of `1.0` vs `1` deliberate?

**Comment by zeptin** [comment]
Looks like the CI doesn't like the new versioning package (this doesn't happen when I build locally):

`error NU1100: Unable to resolve 'Microsoft.AspNetCore.Mvc.Versioning (>= 3.0.0)' for '.NETCoreApp,Version=v2.1'.`

File: src/Stratis.Bitcoin.Features.SmartContracts/ReflectionExecutor/Controllers/SmartContractsController.cs (hunk 1 above)
Comment (codingupastorm) [reply src/Stratis.Bitcoin.Features.SmartContracts/ReflectionExecutor/Controllers/SmartContractsController.cs]:
```suggestion
    [ApiVersion("1")]
```

File: src/Stratis.Bitcoin.Features.Api/ConfigureSwaggerOptions.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    /// Configures the Swagger generation options.
+    /// </summary>
+    /// <remarks>This allows API versioning to define a Swagger document per API version after the
+    /// <see cref="IApiVersionDescriptionProvider"/> service has been resolved from the service container.
+    /// Adapted from https://github.com/microsoft/aspnet-api-versioning/blob/master/samples/aspnetcore/SwaggerSample/ConfigureSwaggerOptions.cs.
+    /// </remarks>
+    public class ConfigureSwaggerOptions : IConfigureOptions<SwaggerGenOptions>
+    {
+        private const string ApiXmlFilename = "Stratis.Bitcoin.Api.xml";
+        private const string WalletXmlFilename = "Stratis.Bitcoin.LightWallet.xml";
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Api/ConfigureSwaggerOptions.cs]: Why is this the llight wallet?

File: src/Stratis.Bitcoin.Features.Api/Startup.cs [hunk 3]
```diff
@@ -69,32 +71,37 @@ public void ConfigureServices(IServiceCollection services)
...
+            services.AddApiVersioning(options =>
             {
-                setup.SwaggerDoc("v1", new Info { Title = "Stratis.Bitcoin.Api", Version = "v1" });
+                // Our versions are configured to be set via URL path, no need to read from querystring etc.
+                options.ApiVersionReader = new UrlSegmentApiVersionReader();
 
-                //Set the comments path for the swagger json and ui.
-                string basePath = PlatformServices.Default.Application.ApplicationBasePath;
-                string apiXmlPath = Path.Combine(basePath, "Stratis.Bitcoin.Api.xml");
-                string walletXmlPath = Path.Combine(basePath, "Stratis.Bitcoin.LightWallet.xml");
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Api/Startup.cs]: Ah I see it was copied from here... Still, do you what the significance is?

Comment (codingupastorm) [reply src/Stratis.Bitcoin.Features.Api/Startup.cs]: I have no idea. Just copied it over. My guess is at some point previously there was documentation missing from some of the APIs? 

We could experiment with removing it, but probably best for a different PR.

Comment (MithrilMan) [reply src/Stratis.Bitcoin.Features.Api/Startup.cs]: from what I know swagger can read up generated xml doc to fill up its API description using comment, I don't know if we are using it properly or not (and of course the build have to generate these xml)

**Comment by codingupastorm** [comment]
@zeptin @fassadlr can I get you guys to once-over this again? Finally all the tests etc are fixed. Only change I made is in the RPC stuff, you can see in the last commit :)

File: src/Stratis.Bitcoin.FullNode.sln [hunk 4]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.29020.237
+# Visual Studio 15
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.FullNode.sln]: @codingupastorm what are these changes? Are they needed?

Comment (codingupastorm) [reply src/Stratis.Bitcoin.FullNode.sln]: woops I maybe used the wrong VS.

**Pull request merged**